NRLMMD-GEOIPS/geoips
# Add reader_kwargs command line option

# Reviewer Checklist

Testing out the new reviewer checklist.  I'll let someone else decide if we need to add unit tests.... 

* [x] Existing tests pass (integration_tests/full_test.sh, test_all.sh, pytests as appropriate.
      Includes linting checks and documentation build)
* [x] Appropriate unit tests added and pass for new/modified functionality
  * [x] No unit tests required (covered via integration tests)
* [x] Appropriate integration tests added and pass for new/modified functionality
* [x] Appropriate documentation added for new/modified functionality
* [x] Appropriate release notes added for new/modified functionality

20230918 I merged 4 branches into a single local "test_changes" branch and successfully ran "full_test.sh" on all:

https://github.com/NRLMMD-GEOIPS/geoips/pull/319
https://github.com/NRLMMD-GEOIPS/geoips/pull/320
https://github.com/NRLMMD-GEOIPS/geoips/pull/316
https://github.com/NRLMMD-GEOIPS/geoips/pull/314

Note the print statement PR required a single update (log->LOG in AHI reader), otherwise these all passed without modification.

I am good merging all of these PRs.

Package: geoips_full
Total run time: 2416 seconds
Number data types run: 47
Number data types failed: 0

https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues
fixes NRLMMD-GEOIPS/geoips#313


# Testing Instructions
We probably want to update these tests to fully implement unit tests, but this will test that it works for the time being:

```
single source command line reader_kwargs test:
./tests/scripts/amsr2.tc.89H-Physical.imagery_annotated.sh

single source defaults test:
./tests/scripts/amsr2_ocean.tc.windspeed.imagery_clean.sh

config based defaults test:
./tests/scripts/amsr2.config_based_overlay_output.sh

config based YAML defined reader_kwargs test:
./tests/scripts/amsr2.config_based_overlay_output_low_memory.sh

config based command line reader_kwargs AND fuse_reader_kwargs test:
./tests/scripts/amsr2.config_based_no_compare.sh
```

full_test.sh should still pass to ensure nothing broke.

# Summary
Add reader_kwargs command line argument
---------------------------------------

Ensure arbitrary keyword arguments can be passed to specified reader.

::

    geoips/commandline/args.py
    geoips/plugins/modules/procflows/single_source.py
    geoips/plugins/modules/procflows/config_based.py

# Output

```
full_test.sh (template_basic_plugin was on wrong branch, only failures): 

Package: geoips_full
Total run time: 2642 seconds
Number data types run: 47
Number data types failed: 2
```

Add "test_arg" to amsr2 reader:
```
+def call(fnames, metadata_only=False, chans=None, area_def=None, self_register=False,
+         test_arg="AMSR2 Default Test Arg"):
```

Including non-existent argument with single source procflow:
```
+++ b/tests/scripts/amsr2.tc.89H-Physical.imagery_annotated.sh
+          --reader_kwargs \
+            '{"test_arg": "hi there", "test_arg_2": "testing"}' \

TypeError: call() got an unexpected keyword argument 'test_arg_2'
```

Include correct argument with single source procflow:
```
+++ b/tests/scripts/amsr2.tc.89H-Physical.imagery_annotated.sh
+          --reader_kwargs \
+            '{"test_arg": "AMSR2 single source command line reader test_arg"}' \
05_111009 single_source.py:1393 INTERACTIVE: Reading full dataset with reader 'amsr2_netcdf'...
05_111009 amsr2_netcdf.py:327  INTERACTIVE: AMSR2 reader test_arg: AMSR2 single source command line reader test_arg
```

Use default with single source procflow:
```
./tests/scripts/amsr2_ocean.tc.windspeed.imagery_clean.sh

05_102745 single_source.py:1393 INTERACTIVE: Reading full dataset with r
eader 'amsr2_netcdf'...
05_102745 amsr2_netcdf.py:327  INTERACTIVE: Test arg: AMSR2 Default Test Arg
```

Use bg_reader_kwargs and reader_kwargs command line options for config-based procflow:
```
./tests/scripts/amsr2.config_based_no_compare.sh

05_113408 config_based.py:860  INTERACTIVE: Reading background datasets using reader 'ahi_hsd'...
05_113408      ahi_hsd.py:983  INTERACTIVE: AHI reader test_arg: Command line config-based ahi test arg
05_113408 config_based.py:878  INTERACTIVE: Reading metadata from datasets using reader 'amsr2_netcdf'...
05_113408 amsr2_netcdf.py:333  INTERACTIVE: AMSR2 reader test_arg: Command line config-based amsr2 test arg
05_113409 config_based.py:904  INTERACTIVE: Reading full dataset using reader 'amsr2_netcdf'...
05_113409 amsr2_netcdf.py:333  INTERACTIVE: AMSR2 reader test_arg: Command line config-based amsr2 test arg
```

Use reader_kwargs defined in YAML, default bg_reader_kwargs:
```
./tests/scripts/amsr2.config_based_overlay_output_low_memory.sh

 reader_name: amsr2_netcdf
 reader_kwargs:
-  test_arg: "Config based reader_kwargs test_arg"
+  test_arg: "Config based YAML reader_kwargs test_arg"

05_114309 config_based.py:860  INTERACTIVE: Reading background datasets using reader 'ahi_hsd'...
05_114309      ahi_hsd.py:983  INTERACTIVE: AHI reader test_arg: AHI Default Test Arg
05_114309 config_based.py:878  INTERACTIVE: Reading metadata from datasets using reader 'amsr2_netcdf'...
05_114309 amsr2_netcdf.py:333  INTERACTIVE: AMSR2 reader test_arg: Config based YAML reader_kwargs test_arg
05_114309 config_based.py:904  INTERACTIVE: Reading full dataset using reader 'amsr2_netcdf'...
05_114309 amsr2_netcdf.py:333  INTERACTIVE: AMSR2 reader test_arg: Config based YAML reader_kwargs test_arg
```

All defaults for config-based:
```
./tests/scripts/amsr2.config_based_overlay_output.sh

05_114433 config_based.py:860  INTERACTIVE: Reading background datasets using reader 'ahi_hsd'...
05_114433      ahi_hsd.py:983  INTERACTIVE: AHI reader test_arg: AHI Default Test Arg
05_114433 config_based.py:878  INTERACTIVE: Reading metadata from datasets using reader 'amsr2_netcdf'...
05_114433 amsr2_netcdf.py:333  INTERACTIVE: AMSR2 reader test_arg: AMSR2 Default Test Arg
05_114434 config_based.py:904  INTERACTIVE: Reading full dataset using reader 'amsr2_netcdf'...
05_114434 amsr2_netcdf.py:333  INTERACTIVE: AMSR2 reader test_arg: AMSR2 Default Test Arg
```
